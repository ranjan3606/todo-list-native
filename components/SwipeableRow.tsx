import React, { useRef, forwardRef, useImperativeHandle } from 'react';
import { Animated, StyleSheet, Text, View, I18nManager, Pressable } from 'react-native';
import { Swipeable } from 'react-native-gesture-handler';
import { FontAwesome } from '@expo/vector-icons';
import { Colors } from '@/constants/Colors';
import { useActualColorScheme } from '@/services/theme';
import { useTranslation } from '@/i18n';

interface SwipeableRowProps {
  children: React.ReactNode;
  onDelete: () => void;
  onComplete?: () => void;
  onSnooze?: () => void;
  isCompleted?: boolean;
}

// Define a ref interface for exposing methods
export interface SwipeableRowHandle {
  close: () => void;
}

// Use forwardRef to properly handle refs
export const SwipeableRow = forwardRef<SwipeableRowHandle, SwipeableRowProps>(({ 
  children, 
  onDelete, 
  onComplete,
  onSnooze,
  isCompleted = false
}, ref) => {
  const swipeableRef = useRef<Swipeable>(null);
  const colorScheme = useActualColorScheme();
  const { t } = useTranslation();
  
  // Expose methods via ref
  useImperativeHandle(ref, () => ({
    close: () => {
      swipeableRef.current?.close();
    }
  }));

  const renderRightAction = (
    text: string,
    color: string,
    icon: string, 
    x: number,
    progress: Animated.AnimatedInterpolation<number>,
    action: () => void
  ) => {
    const trans = progress.interpolate({
      inputRange: [0, 1],
      outputRange: [x, 0],
    });
    
    const pressHandler = () => {
      swipeableRef.current?.close();
      action();
    };

    return (
      <Animated.View style={{ flex: 1, transform: [{ translateX: trans }] }}>
        <Pressable
          style={({ pressed }) => [
            styles.rightAction, 
            { backgroundColor: color },
            pressed && { opacity: 0.7 }
          ]}
          onPress={pressHandler}>
          <FontAwesome name={icon} size={20} color="white" style={styles.actionIcon} />
          <Text style={styles.actionText}>{text}</Text>
        </Pressable>
      </Animated.View>
    );
  };

  const renderRightActions = (progress: Animated.AnimatedInterpolation<number>) => (
    <View style={{ width: 80, flexDirection: I18nManager.isRTL ? 'row-reverse' : 'row' }}>
      {renderRightAction(t('common.delete'), Colors.danger, 'trash', 80, progress, onDelete)}
    </View>
  );

  const renderLeftAction = (
    text: string,
    color: string,
    icon: string,
    x: number,
    progress: Animated.AnimatedInterpolation<number>,
    action: () => void
  ) => {
    const trans = progress.interpolate({
      inputRange: [0, 1],
      outputRange: [x, 0],
    });
    
    const pressHandler = () => {
      swipeableRef.current?.close();
      action();
    };

    return (
      <Animated.View style={{ flex: 1, transform: [{ translateX: trans }] }}>
        <Pressable
          style={({ pressed }) => [
            styles.leftAction, 
            { backgroundColor: color },
            pressed && { opacity: 0.7 }
          ]}
          onPress={pressHandler}>
          <FontAwesome name={icon} size={20} color="white" style={styles.actionIcon} />
          <Text style={styles.actionText}>{text}</Text>
        </Pressable>
      </Animated.View>
    );
  };

  const renderLeftActions = (progress: Animated.AnimatedInterpolation<number>) => {
    if (!onComplete && !onSnooze) return null;
    
    const actionWidth = 100;
    const actions = [];
    
    if (onSnooze) {
      actions.push(
        renderLeftAction(
          t('task.snooze'), 
          Colors.info, 
          'clock-o', 
          -actionWidth, 
          progress, 
          onSnooze
        )
      );
    }
    
    if (onComplete) {
      actions.push(
        renderLeftAction(
          isCompleted ? t('task.undo') : t('task.done'), 
          Colors.success, 
          isCompleted ? 'undo' : 'check', 
          onSnooze ? -2 * actionWidth : -actionWidth, 
          progress, 
          onComplete
        )
      );
    }
    
    return (
      <View style={{ width: actions.length * actionWidth, flexDirection: I18nManager.isRTL ? 'row-reverse' : 'row' }}>
        {actions.map((action, index) => (
          <React.Fragment key={`action-${index}`}>
            {action}
          </React.Fragment>
        ))}
      </View>
    );
  };

  return (
    <View style={[styles.container, { backgroundColor: Colors[colorScheme].background }]}>
      <Swipeable
        ref={swipeableRef}
        friction={2}
        leftThreshold={30}
        rightThreshold={40}
        renderRightActions={renderRightActions}
        renderLeftActions={renderLeftActions}>
        {children}
      </Swipeable>
    </View>
  );
});

const styles = StyleSheet.create({
  container: {
    backgroundColor: 'white',
  },
  leftAction: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    flexDirection: 'row',
  },
  actionIcon: {
    marginRight: 8,
  },
  actionText: {
    color: 'white',
    fontSize: 14,
    fontWeight: '500',
  },
  rightAction: {
    alignItems: 'center',
    flex: 1,
    justifyContent: 'center',
    flexDirection: 'row',
  },
});
