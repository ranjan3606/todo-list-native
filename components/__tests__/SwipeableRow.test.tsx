import React from 'react';
import { render, fireEvent, within } from '@testing-library/react-native';
import { Text, View } from 'react-native';
import { SwipeableRow } from '../SwipeableRow';

// Mock @expo/vector-icons to prevent native font loading issues
jest.mock('@expo/vector-icons', () => {
  const React = require('react');
  const { Text } = require('react-native');
  return {
    FontAwesome: function MockFontAwesome(props) {
      return React.createElement(
        Text,
        { ...props, testID: `icon-${props.name}` },
        props.name
      );
    }
  };
});

jest.mock('@react-native-async-storage/async-storage', () => ({
  getItem: jest.fn(() => Promise.resolve(null)),
  setItem: jest.fn(() => Promise.resolve(null)),
  removeItem: jest.fn(() => Promise.resolve(null)),
  clear: jest.fn(() => Promise.resolve(null)),
  getAllKeys: jest.fn(() => Promise.resolve([])),
  multiGet: jest.fn(() => Promise.resolve([])),
  multiSet: jest.fn(() => Promise.resolve(null)),
  multiRemove: jest.fn(() => Promise.resolve(null)),
}));
// Mock react-native-gesture-handler
jest.mock('react-native-gesture-handler', () => {
  const View = require('react-native').View;
  const Animated = require('react-native').Animated;
  return {
    Swipeable: ({ renderLeftActions, renderRightActions, children }) => {
      const left = renderLeftActions ? renderLeftActions(new Animated.Value(1)) : null;
      const right = renderRightActions ? renderRightActions(new Animated.Value(1)) : null;
      return (
        <View>
          {left && <View testID="left-actions">{left}</View>}
          <View testID="swipeable-content">{children}</View>
          {right && <View testID="right-actions">{right}</View>}
        </View>
      );
    },
  };
});

jest.mock('@/constants/Colors', () => ({
  Colors: {
    light: { background: '#fff' },
    dark: { background: '#000' },
    danger: 'red',
    success: 'green',
    info: 'blue',
  }
}));

jest.mock('@/services/theme', () => ({
  useActualColorScheme: jest.fn().mockReturnValue('light'),
}));

jest.mock('@/i18n', () => ({
  useTranslation: () => ({
    t: (key: string) => {
      const translations = {
        'common.delete': 'Delete',
        'task.done': 'Done',
        'task.undo': 'Undo',
        'task.snooze': 'Snooze'
      };
      return translations[key] || key;
    }
  })
}));

describe('SwipeableRow', () => {
  const mockDelete = jest.fn();
  const mockComplete = jest.fn();
  const mockSnooze = jest.fn();
  
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders correctly (snapshot)', () => {
    const { toJSON } = render(
      <SwipeableRow onDelete={mockDelete}>
        <Text>Content</Text>
      </SwipeableRow>
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders child content', () => {
    const { getByText } = render(
      <SwipeableRow onDelete={mockDelete}>
        <Text>Test Child Content</Text>
      </SwipeableRow>
    );
    
    expect(getByText('Test Child Content')).toBeTruthy();
  });

  it('renders delete action on the right side', () => {
    const { getByTestId } = render(
      <SwipeableRow onDelete={mockDelete}>
        <Text>Content</Text>
      </SwipeableRow>
    );
    
    // Right actions should be rendered
    const rightActions = getByTestId('right-actions');
    expect(rightActions).toBeTruthy();
  });

  it('renders complete and snooze actions on the left side when provided', () => {
    const { getByTestId } = render(
      <SwipeableRow onDelete={mockDelete} onComplete={mockComplete} onSnooze={mockSnooze}>
        <Text>Content</Text>
      </SwipeableRow>
    );
    
    // Left actions should be rendered
    const leftActions = getByTestId('left-actions');
    expect(leftActions).toBeTruthy();
  });

  it('does not render left actions when onComplete and onSnooze are not provided', () => {
    const { queryByTestId } = render(
      <SwipeableRow onDelete={mockDelete}>
        <Text>Content</Text>
      </SwipeableRow>
    );
    
    // Left actions should not be rendered
    expect(queryByTestId('left-actions')).toBeNull();
  });

  it('uses "Undo" text for complete action when task is already completed', () => {
    const { getByTestId } = render(
      <SwipeableRow onDelete={mockDelete} onComplete={mockComplete} isCompleted={true}>
        <Text>Content</Text>
      </SwipeableRow>
    );
    
    const leftActions = getByTestId('left-actions');
    expect(within(leftActions).getByText('Undo')).toBeTruthy();
  });

  it('uses ref to expose close method', () => {
    const ref = React.createRef();
    
    render(
      <SwipeableRow ref={ref} onDelete={mockDelete}>
        <Text>Content</Text>
      </SwipeableRow>
    );
    
    // The ref should have a close method
    expect(ref.current).toBeTruthy();
    expect(typeof ref.current.close).toBe('function');
  });
});