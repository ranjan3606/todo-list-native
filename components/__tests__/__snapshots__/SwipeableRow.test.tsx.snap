// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SwipeableRow renders correctly (snapshot) 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "white",
      },
      {
        "backgroundColor": "#fff",
      },
    ]
  }
>
  <View>
    <View
      testID="swipeable-content"
    >
      <Text>
        Content
      </Text>
    </View>
    <View
      testID="right-actions"
    >
      <View
        style={
          {
            "flexDirection": "row",
            "width": 80,
          }
        }
      >
        <View
          collapsable={false}
          style={
            {
              "flex": 1,
              "transform": [
                {
                  "translateX": 0,
                },
              ],
            }
          }
        >
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              [
                {
                  "alignItems": "center",
                  "flex": 1,
                  "flexDirection": "row",
                  "justifyContent": "center",
                },
                {
                  "backgroundColor": "red",
                },
                false,
              ]
            }
          >
            <Text
              color="white"
              name="trash"
              size={20}
              style={
                {
                  "marginRight": 8,
                }
              }
              testID="icon-trash"
            >
              trash
            </Text>
            <Text
              style={
                {
                  "color": "white",
                  "fontSize": 14,
                  "fontWeight": "500",
                }
              }
            >
              Delete
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
